File: services/api/src/agas_api/athletic_dashboard.py
from __future__ import annotations

from collections import Counter
from datetime import UTC, datetime
from typing import Literal
from uuid import UUID

from agas_domain import CapabilityDomain, CapabilityEstimate, Confidence
from agas_domain.persistence.repository import DomainRepository
from pydantic import BaseModel, ConfigDict, JsonValue
from sqlalchemy.orm import Session

CapabilitySeriesStatus = Literal["current", "stale"]
CapabilityDomainStatus = Literal["not_estimated", "current", "stale", "mixed"]


class CapabilitySeriesProjection(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    estimate_id: UUID
    kind: Literal["derived"] = "derived"
    estimate_scope: str
    estimate: JsonValue
    unit_or_scale: str
    confidence: Confidence
    status: CapabilitySeriesStatus
    calculation_method: str
    source_observation_ids: tuple[UUID, ...]
    estimated_at: datetime
    valid_until: datetime | None
    rule_version: str
    historical_estimate_count: int


class CapabilityDomainProjection(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    domain: CapabilityDomain
    status: CapabilityDomainStatus
    latest_estimates: tuple[CapabilitySeriesProjection, ...]
    historical_estimate_count: int


class AthleticDashboardProjection(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    athlete_id: UUID
    athlete_display_name: str
    as_of: datetime
    domains: tuple[CapabilityDomainProjection, ...]
    estimated_domain_count: int
    unestimated_domain_count: int
    projection_version: str = "athletic-dashboard-projection@1.0.0"


class AthleticDashboardNotFoundError(LookupError):
    pass


def get_athletic_dashboard_projection(
    session: Session, athlete_id: UUID, as_of: datetime | None = None
) -> AthleticDashboardProjection:
    repository = DomainRepository(session)
    athlete = repository.get_athlete(athlete_id)
    if athlete is None:
        raise AthleticDashboardNotFoundError("athlete does not exist")
    instant = as_of or datetime.now(UTC)
    if instant.tzinfo is None or instant.utcoffset() is None:
        raise ValueError("athletic dashboard time must include a timezone")

    visible_estimates = tuple(
        estimate
        for estimate in repository.list_capability_estimates(athlete_id)
        if estimate.estimated_at <= instant
    )
    series_counts = Counter(
        (estimate.domain, estimate.estimate_scope, estimate.unit_or_scale)
        for estimate in visible_estimates
    )
    latest_by_series: dict[tuple[CapabilityDomain, str, str], CapabilityEstimate] = {}
    for estimate in visible_estimates:
        key = (estimate.domain, estimate.estimate_scope, estimate.unit_or_scale)
        latest_by_series.setdefault(key, estimate)

    domains = []
    for domain in CapabilityDomain:
        domain_keys = sorted(
            (key for key in latest_by_series if key[0] is domain),
            key=lambda key: (key[1], key[2]),
        )
        domain_series = tuple(
            CapabilitySeriesProjection(
                estimate_id=estimate.id,
                estimate_scope=estimate.estimate_scope,
                estimate=estimate.estimate,
                unit_or_scale=estimate.unit_or_scale,
                confidence=estimate.confidence,
                status=(
                    "stale"
                    if estimate.valid_until is not None and estimate.valid_until <= instant
                    else "current"
                ),
                calculation_method=estimate.calculation_method,
                source_observation_ids=estimate.source_observation_ids,
                estimated_at=estimate.estimated_at,
                valid_until=estimate.valid_until,
                rule_version=estimate.rule_version,
                historical_estimate_count=series_counts[
                    (estimate.domain, estimate.estimate_scope, estimate.unit_or_scale)
                ],
            )
            for key in domain_keys
            for estimate in (latest_by_series[key],)
        )
        statuses = {item.status for item in domain_series}
        if not statuses:
            domain_status: CapabilityDomainStatus = "not_estimated"
        elif len(statuses) > 1:
            domain_status = "mixed"
        elif "stale" in statuses:
            domain_status = "stale"
        else:
            domain_status = "current"
        domains.append(
            CapabilityDomainProjection(
                domain=domain,
                status=domain_status,
                latest_estimates=domain_series,
                historical_estimate_count=sum(
                    count for key, count in series_counts.items() if key[0] is domain
                ),
            )
        )

    estimated_domain_count = sum(item.status != "not_estimated" for item in domains)
    return AthleticDashboardProjection(
        athlete_id=athlete.id,
        athlete_display_name=athlete.display_name,
        as_of=instant,
        domains=tuple(domains),
        estimated_domain_count=estimated_domain_count,
        unestimated_domain_count=len(domains) - estimated_domain_count,
    )

File: apps/web/lib/athletic-dashboard.ts
import { authorizedHeaders } from "./identity";
import type { Confidence } from "./current-week";

export type CapabilitySeriesStatus = "current" | "stale";
export type CapabilityDomainStatus = "not_estimated" | "current" | "stale" | "mixed";

export interface CapabilitySeriesProjection {
  estimate_id: string;
  kind: "derived";
  estimate_scope: string;
  estimate: unknown;
  unit_or_scale: string;
  confidence: Confidence;
  status: CapabilitySeriesStatus;
  calculation_method: string;
  source_observation_ids: string[];
  estimated_at: string;
  valid_until: string | null;
  rule_version: string;
  historical_estimate_count: number;
}

export interface CapabilityDomainProjection {
  domain: string;
  status: CapabilityDomainStatus;
  latest_estimates: CapabilitySeriesProjection[];
  historical_estimate_count: number;
}

export interface AthleticDashboardProjection {
  athlete_id: string;
  athlete_display_name: string;
  as_of: string;
  domains: CapabilityDomainProjection[];
  estimated_domain_count: number;
  unestimated_domain_count: number;
  projection_version: string;
}

export class AthleticDashboardRequestError extends Error {
  constructor(
    message: string,
    readonly status: number,
  ) {
    super(message);
    this.name = "AthleticDashboardRequestError";
  }
}

export async function fetchAthleticDashboard(
  apiBaseUrl: string,
  athleteId: string,
  fetcher: typeof fetch = fetch,
): Promise<AthleticDashboardProjection> {
  const response = await fetcher(
    `${apiBaseUrl.replace(/\/$/, "")}/v1/athletes/${encodeURIComponent(athleteId)}/dashboard`,
    { headers: authorizedHeaders({ Accept: "application/json" }) },
  );
  if (!response.ok) {
    const payload = (await response.json().catch(() => null)) as { detail?: string } | null;
    throw new AthleticDashboardRequestError(
      payload?.detail ?? `Athletic dashboard request failed with ${response.status}.`,
      response.status,
    );
  }
  return (await response.json()) as AthleticDashboardProjection;
}

export function capabilityDomainLabel(domain: string): string {
  return domain
    .split("_")
    .map((part) => `${part.slice(0, 1).toUpperCase()}${part.slice(1)}`)
    .join(" ");
}

export function capabilityValueLabel(value: unknown, unitOrScale: string): string {
  const rendered =
    typeof value === "number" || typeof value === "string" ? String(value) : JSON.stringify(value);
  return `${rendered} ${unitOrScale}`.trim();
}
